src/main/scala/cs332/distributedsorting/master/Master.scala

```diff
@@ -45,12 +66,68 @@ class Master(executionContext: ExecutionContext) { self =>
     }
   }
 
-  private class GreeterImpl extends GreeterGrpc.Greeter {
-    override def sayHello(req: HelloRequest) = {
-      Master.logger.info("Hello from " + req.name)
-      val reply = HelloReply(message = "Hello " + req.name)
+
+  private def addData(data : String,ipAddress : String) : Boolean = {
+    this.synchronized({
+      if (slaves.filter(x=>x.ip == ipAddress).isEmpty)
+        System.out.println("we receive data from a client we did not register")
+        return false 
+      count +=1
+      this.data :+ data.getBytes().grouped(10).toList
+      if(count == numClient)
+        System.out.println("we receive all the data")
+      return true
+    })
+  }
+
+  private def createPartition(): Map[String, (Array[Byte], Array[Byte])] = {
+    val mindata = data.min(KeyOrdering)
+    val maxdata = data.max(KeyOrdering)
+    val range = div(count, sub(maxdata,mindata))
+    var loop  = 0
+    for (slave <- slaves ){
+      partition += (slave.toString ->(add(mindata, mult(loop,range)), add(mindata, mult(loop+1,range))))
+      loop+=1
+    }
+    return partition
+  }
```

Comment:
I think that we have to consider the distribution of sampled key for deciding the key ranges.

Every key is 10 byte long, so it means a key can range over from 0 to (2^8)^10 = 2^80. So we should make our application to handle all the keys between 0 to 2^80 for sorting.

If we have `n` slaves, we should count the length of all sampled keys and divide them into `n` equal length sorted chunks. The last key of each chunk will be the upper bound for each `n` partition.
Also there are exceptional cases. The first partition must have lowerbound of 0 and the last partition must have upperbound of 2^80.